python/hail/docs/tutorials/08-joins.ipynb

```diff
@@ -0,0 +1,377 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "## Joins\n",
+    "\n",
+    "The movie dataset comes in multiple parts.  It is very natural to ask questions about the relationship between the parts.\n",
+    "\n",
+    "- What is the mean rating per genre?\n",
+    "- What is the favorite movie for each occupation?\n",
+    "- What genres are most preferred by women vs men?\n",
+    " \n",
+    "Joins let us combine multiple datasets together to answer joint questions."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "import hail as hl\n",
+    "import matplotlib.pyplot as plt\n",
+    "import seaborn\n",
+    "%matplotlib inline\n",
+    "\n",
+    "hl.utils.get_movie_lens('data/')\n",
+    "\n",
+    "users = hl.read_table('data/users.ht')\n",
+    "movies = hl.read_table('data/movies.ht')\n",
+    "ratings = hl.read_table('data/ratings.ht')"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "To understand joins in Hail, we need to revisit one of the crucial properties of `Table`s: the key.\n",
+    "\n",
+    "A `Table` has an ordered list of fields known as the key.  The key is shown by `describe`."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "users.describe()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "`key` is a struct expression of all of the key fields."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "slideshow": {
+     "slide_type": "fragment"
+    }
+   },
+   "outputs": [],
+   "source": [
+    "users.key.describe()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "slideshow": {
+     "slide_type": "slide"
+    }
+   },
+   "source": [
+    "Hail makes no assumptions on keys.  In particular, the keys need not be unique or non-missing, although in many applications they will be both.\n"
```

Comment:
Hail makes no assumptions on key values.  In particular, they need not be unique or non-missing,